Add `clientClaims` request parameter for confidential client (#8686)
AzureAD/microsoft-authentication-library-for-js: lib/msal-node/docs/request.md

The assertion callback context (`ClientAssertionConfig`) includes `fmiPath` so that context-aware assertion callbacks can use it to acquire the correct credential for multi-leg agent flows.

### Client-originated claims (`clientClaims`)

All confidential client flows — `acquireTokenByClientCredential`, `acquireTokenOnBehalfOf`, and `acquireTokenByUserFederatedIdentityCredential` — accept an optional `clientClaims` parameter. It forwards **client-originated** claims to the token endpoint, sent as the `claims` parameter in the request body. The value is forwarded as-is — MSAL does not restrict which claim keys you send. When both `claims` and `clientClaims` are present they are **deep-merged**, with `clientClaims` taking precedence on conflicting keys (nested objects are merged recursively rather than replaced).

> Unlike `claims` (a server-issued challenge, which **bypasses** the token cache), `clientClaims` does **not** bypass the cache. Tokens are cached and the cache entry is **keyed on the `clientClaims` value** (using the same extended cache-key hash as `fmiPath`): identical values are served from cache, while different values produce separate cache entries. Because the entry is keyed on the value, send the **same `clientClaims` value on every request** for which you want the cached token to be reused — a different value (or omitting it) produces a separate cache entry and a new network call. Use stable, non-dynamic values to avoid unbounded cache growth. Empty, whitespace-only, or empty-object (`{}`) values are ignored.

Reviewer: Curious as to why we're making client claims behave differently from the base claims? Do we not have the same caching concerns we had with the existing claims request?

Author: Intentional - server `claims` is an ephemeral challenge (e.g. CAE) that must hit the network, so it bypasses the cache read; `clientClaims` is a stable, client-supplied attribute sent on every call, so instead of bypassing we partition the cache key on its value - a token minted with clientClaims=X is only ever served to a request with clientClaims=X. Mirrors msal-dotnet. Added a clarifying sentence to the docs.

Reviewer 2: What if a client makes calls with a new claim per request? We have seen users leverage this as a mode for one time use tokens (which they think enhance security sometimes). Will the cache not overflow?

> > Note: `acquireTokenByUserFederatedIdentityCredential` always calls the network, so `clientClaims` is forwarded on every request but is not cached by that method.

Author: Reworded in ffa3d4896 to avoid implying tokens aren't stored: FIC forwards clientClaims every call and tokens are still cached, just not partitioned by clientClaims.

> Unlike `claims` (a server-issued challenge, which **bypasses** the token cache), `clientClaims` does **not** bypass the cache. This difference is intentional: server `claims` is an _ephemeral_ challenge (for example, a CAE claims challenge) that must be satisfied by a fresh network call, whereas `clientClaims` is a _stable, client-supplied_ attribute (for example, a network-perimeter id) sent on every request — so instead of bypassing the cache it **partitions** it by value. Partitioning addresses the same cross-request contamination concern that motivates the `claims` bypass: a token minted with a given `clientClaims` value is only ever served to a request presenting that same value. Tokens are cached and the cache entry is **keyed on the `clientClaims` value** (using the same extended cache-key hash as `fmiPath`): identical values are served from cache, while different values produce separate cache entries. Because the entry is keyed on the value, send the **same `clientClaims` value on every request** for which you want the cached token to be reused — a different value (or omitting it) produces a separate cache entry and a new network call. Use stable, non-dynamic values to avoid unbounded cache growth. Empty, whitespace-only, or empty-object (`{}`) values are ignored.

Reviewer: Make this a bit more concise and expectations more explicit - "can be" vs "must be"
```suggestion
Unlike `claims` (a server-issued challenge, which **bypasses** the token cache), `clientClaims` does **not** bypass the cache. Server `claims` can be an _ephemeral_ challenge (for example, a CAE claims challenge) that must be satisfied by a fresh network call, whereas `clientClaims` must be a _stable, client-supplied_ attribute (for example, a network-perimeter id) sent on every request — so instead of bypassing the cache it **partitions** it by value. Because the entry is keyed on the value, send the **same `clientClaims` value on every request** for which you want the cached token to be reused — a different value (or omitting it) produces a separate cache entry and a new network call. You must pass stable, non-dynamic values to avoid unbounded cache growth.
```

Author: done
